### DISCUSSION TITLE
Easy dark mode switching from UI

### DISCUSSION THREAD

**Discussion by smojef** [zauberzeug/nicegui#342] [Ideas / Feature Requests]
Here is a feature request.  Looking at my Hello World #339 and the bugfix #311, it would be desirable to expose an easy way to change dark mode.

Things to consider: 
- Ideally we want to bind the dark mode to an element like: `ui.switch('Dark').bind_value(ui, 'dark_mode')`
- The initial dark mode can be True, False or None (which is determined later from the browser, I guess).
- The tricky part is to handle the initial state of `ui.dark_mode` (while creating the UI and binding to it) until `ui.run()` is called.

**Comment by pythonormen** [comment]
Is there still no possible to replace/override the ui.run() value? 
Thank you

**Comment by falkoschindler** [comment]
@pythonormen No, we haven't started implementing this feature. But for switching the dark mode programmatically, I just came up with the following sketch based on the JS code in index.html:

```py
async def make_dark():
    await ui.run_javascript('''
        Quasar.Dark.set(true);
        tailwind.config.darkMode = "class";
        document.body.classList.add("dark");
    ''', respond=False)

async def make_light():
    await ui.run_javascript('''
        Quasar.Dark.set(false);
        tailwind.config.darkMode = "class"
        document.body.classList.remove("dark");
    ''', respond=False)

async def make_auto():
    await ui.run_javascript('''
        Quasar.Dark.set("auto");
        tailwind.config.darkMode = "media";
    ''', respond=False)

ui.button('Make dark', on_click=make_dark)
ui.button('Make light', on_click=make_light)
ui.button('Make auto', on_click=make_auto)
```

This could serve as a starting point for an implementation with a neat API. Pull requests are always welcome! :wink:
But maybe we should clarify how the API should look like first.

**Reply by HilkopterBob** [reply]
any updates?

**Reply by falkoschindler** [reply]
No, we're still not sure about a good API.
Have you tried the functions I proposed? What is missing from your point of view?

**Reply by pythonormen** [reply]
Not HilkopterBob but for my use case it works perfectly fine. A question though. How to adjust the colors used for dark mode?

```py
ui.colors(dark='...') 
ui.colors(dark-page='...') 
```
does not seem to work

**Reply by falkoschindler** [reply]
@pythonormen The "dark" color was missing, but has been added in https://github.com/zauberzeug/nicegui/issues/740. We'll be releasing an update with version 1.2.7 shortly.

**Reply by pythonormen** [reply]
<img width="303" alt="image" src="https://user-images.githubusercontent.com/125542430/232059902-93b06341-7008-4fbb-8117-2192c6098b4e.png">

Nice. Isn't dark-page applicable here as well?

**Reply by falkoschindler** [reply]
Hm, what is "dark-page"? I covered the eight colors from the Quasar Color Palette: https://quasar.dev/style/color-palette#brand-colors

**Reply by pythonormen** [reply]
Ok, maybe not relevant here then. Just checked the theme-builder https://quasar.dev/style/theme-builder

**Comment by bobwatcherx** [comment]
Is there a way to detect that we are in dark mode and customize the color. For example, in dark mode, the text turns yellow or the container that was blue in dark mode turns yellow. And when Light mode is back to blue again

**Reply by falkoschindler** [reply]
Here is a function to detect dark mode:

```py
async def detect_dark_mode():
    result = await ui.run_javascript('''
        if (Quasar.Dark.mode == "auto")
            return window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches;
        else
            return Quasar.Dark.mode;
    ''')
    ui.notify(f'Dark mode: {result}')

ui.button('Detect dark mode', on_click=detect_dark_mode)
```

**Comment by falkoschindler** [comment]
I just opened PR https://github.com/zauberzeug/nicegui/pull/783 that proposes a `ui.dark` element for getting, setting and binding the dark mode. Let me know what you think. 🙂
